### PR TITLE
requirement(oci) implement tag retention for oci

### PR DESCRIPTION
**Pull request by bitsf**
requirement(oci) implement tag retention for oci

### CODE REVIEW

File: src/pkg/clients/core/chartv2.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ import (
 	"github.com/goharbor/harbor/src/chartserver"
 )
 
-func (c *client) ListAllCharts(project, repository string) ([]*chartserver.ChartVersion, error) {
+func (c *clientv2) ListAllCharts(project, repository string) ([]*chartserver.ChartVersion, error) {
 	url := c.buildURL(fmt.Sprintf("/api/chartrepo/%s/charts/%s", project, repository))
```
Comment (ywk253100) [inline src/pkg/clients/core/chartv2.go]: The chart related APIs are also bumped up to v2.0 already

File: src/pkg/art/candidate.go [hunk 2]
```diff
@@ -90,16 +90,9 @@ type Candidate struct {
 // Hash code based on the candidate info for differentiation
 func (c *Candidate) Hash() string {
 	if c.Digest == "" {
-		log.Errorf("Lack Digest of Candidate for %s/%s:%s", c.Namespace, c.Repository, c.Tag)
+		log.Errorf("Lack Digest of Candidate for %s/%s:%s", c.Namespace, c.Repository, c.Digest)
```
Comment (steven-zou) [inline src/pkg/art/candidate.go]: please pay attention, `c.Digest` is empty

File: src/pkg/art/selectors/doublestar/selector.go [hunk 3]
```diff
@@ -85,6 +96,32 @@ func (s *selector) Select(artifacts []*art.Candidate) (selected []*art.Candidate
...
+			return false, err
+		}
+		if matched {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (s *selector) tagSelectExclude(artifact *art.Candidate) (selected bool, err error) {
```
Comment (steven-zou) [inline src/pkg/art/selectors/doublestar/selector.go]: This logic seems has problem, the only case to return `false` is all the tags are excluded, right?

File: src/pkg/retention/dep/client.go [hunk 4]
```diff
@@ -101,20 +102,33 @@ func (bc *basicClient) GetCandidates(repository *art.Repository) ([]*art.Candida
 		}
 		for _, image := range images {
```
Comment (steven-zou) [inline src/pkg/retention/dep/client.go]: image == artifact?

File: src/pkg/clients/core/clientv2.go [hunk 5]
```diff
@@ -16,50 +16,49 @@ package core
...
+	ImageClientv2
+	ChartClientv2
 }
 
-// ImageClient defines the methods that an image client should implement
-type ImageClient interface {
-	ListAllImages(project, repository string) ([]*models.TagResp, error)
-	DeleteImage(project, repository, tag string) error
+// ImageClientv2 defines the methods that an image client should implement
+type ImageClientv2 interface {
```
Comment (steven-zou) [inline src/pkg/clients/core/clientv2.go]: ArtifactClient?

File: src/pkg/clients/core/chartv2.go (hunk 1 above)
Comment (bitsf) [reply src/pkg/clients/core/chartv2.go]: OK, will fix this in another PR

File: src/pkg/clients/core/client.go [hunk 6]
```diff
@@ -29,15 +28,15 @@ import (
...
+	ArtifactClient
 	ChartClient
 }
 
-// ImageClient defines the methods that an image client should implement
-type ImageClient interface {
-	ListAllImages(project, repository string) ([]*models.TagResp, error)
-	DeleteImage(project, repository, tag string) error
-	DeleteImageRepository(project, repository string) error
+// ArtifactClient defines the methods that an image client should implement
```
Comment (steven-zou) [inline src/pkg/clients/core/client.go]: comment is inconsistent

**Pull request merged**